### PR TITLE
Introduce ObservedHolidays class

### PR DESCRIPTION
**Pull request by KJhellico**
## Proposed change

Introduce `ObservedHolidays` class for unified observed holidays population.

## Type of change

- [ ] New country/market holidays support (thank you!)
- [ ] Supported country/market holidays update (calendar discrepancy fix, localization)
- [x] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency update (version deprecation/upgrade)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [ ] New feature (new python-holidays functionality in general)

## Checklist

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] This PR is filed against `beta` branch of the repository
- [x] This PR doesn't contain any merge conflicts and has clean commit history
- [x] The code style looks good: `make pre-commit`
- [x] All tests pass locally: `make test`, `make tox` (we strongly encourage adding tests to your code)
- [x] The related [documentation][docs] has been added/updated (check off the box for free if no updates is required)

[contributing-guidelines]: https://github.com/vacanza/python-holidays/blob/beta/CONTRIBUTING.rst
[docs]: https://github.com/vacanza/python-holidays/tree/beta/docs/source

### CODE REVIEW

**Comment by KJhellico** [comment]
Bulgaria and Hong Kong specific observed rules have not been implemented yet.

**Comment by PPsyrius** [comment]
Nice work ❤️

**Review by arkid15r — COMMENTED** [review]
The idea is pretty much clear and aligns perfectly w/ our general refactoring approach. Let's figure out an implementation that would fit our needs best.

This is mostly a structural review of the PR so expect additional round(s) (or just a commit(s) from me) focused on the naming:

File: holidays/countries/guatemala.py [hunk 1]
```diff
@@ -31,26 +38,11 @@ class Guatemala(HolidayBase, ChristianHolidays, InternationalHolidays):
     def __init__(self, *args, **kwargs):
         ChristianHolidays.__init__(self)
         InternationalHolidays.__init__(self)
+        ObservedHolidays.__init__(self, rule=(0, -1, -2, 4, 3, 2, 1))
```
[the remark this answers is not in the document]
Comment (KJhellico) [reply holidays/countries/guatemala.py:41]: "British" `NEAREST_MON` (used in Canada and New Zealand) differs from this: here, Thursday is moved to next Monday, and there, it's moved to previous Monday. :))

[the remark this answers is not in the document]
Comment (KJhellico) [reply holidays/countries/guatemala.py]: Now it's `NEAREST_MON_LATAM`. :)

File: holidays/countries/greece.py [hunk 2]
```diff
@@ -29,12 +27,15 @@ class Greece(HolidayBase, ChristianHolidays, InternationalHolidays):
 
     country = "GR"
     default_language = "el"
+    # %s (Observed).
+    observed_label = "%s (παρατηρήθηκε)"
```
[the remark this answers is not in the document]
Comment (KJhellico) [reply holidays/countries/greece.py]: Done.

File: holidays/groups/observed.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    def _is_observed_applicable(self, dt: date) -> bool:
+        return (not self._begin) or (self._year >= self._begin)
+
+    def _get_observed_date(self, dt: date, rule: ObservedRule) -> date:
+        delta = rule[dt.weekday()]
+        if delta != 0:
+            if delta in {-7, 7}:
+                delta //= 7
+                dt += td(days=delta)
+                while (dt.year == self._year) and (
```
[the remark this answers is not in the document]
Comment (KJhellico) [reply holidays/groups/observed.py]: Done.

File: holidays/groups/observed.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        self._rule = rule
+        self._begin = begin
+
+    def _is_observed_applicable(self, dt: date) -> bool:
+        return (not self._begin) or (self._year >= self._begin)
+
+    def _get_observed_date(self, dt: date, rule: ObservedRule) -> date:
+        delta = rule[dt.weekday()]
+        if delta != 0:
+            if delta in {-7, 7}:
```
[the remark this answers is not in the document]
Comment (KJhellico) [reply holidays/groups/observed.py]: Done.

File: holidays/groups/observed.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        begin: Optional[int] = None,
+    ) -> None:
+        self._rule = rule
+        self._begin = begin
+
+    def _is_observed_applicable(self, dt: date) -> bool:
+        return (not self._begin) or (self._year >= self._begin)
+
+    def _get_observed_date(self, dt: date, rule: ObservedRule) -> date:
+        delta = rule[dt.weekday()]
```
[the remark this answers is not in the document]
Comment (KJhellico) [reply holidays/groups/observed.py]: Done.

File: holidays/groups/observed.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    def __init__(
+        self,
+        rule: ObservedRule,
+        begin: Optional[int] = None,
+    ) -> None:
+        self._rule = rule
+        self._begin = begin
+
+    def _is_observed_applicable(self, dt: date) -> bool:
+        return (not self._begin) or (self._year >= self._begin)
```
[the remark this answers is not in the document]
Comment (KJhellico) [reply holidays/groups/observed.py]: Done.

File: holidays/constants.py [hunk 7]
```diff
@@ -65,3 +65,26 @@
...
+THU_TO_WED_AND_FRI_TO_SAT = (0, 0, 0, -1, 1, 0, 0)
+FRI_TO_THU_AND_SAT_TO_SUN = (0, 0, 0, 0, -1, 1, 0)
+FRI_TO_NEXTWORK = (0, 0, 0, 0, 7, 0, 0)
+SAT_TO_NEXTWORK = (0, 0, 0, 0, 0, 7, 0)
+SUN_TO_NEXTWORK = (0, 0, 0, 0, 0, 0, 7)
+WEEKEND_TO_NEXTWORK = (0, 0, 0, 0, 0, 7, 7)
+WEEKEND_TO_PREV_NEXT_WORK = (0, 0, 0, 0, 0, -7, 7)
+THU_FRI_TO_NEXTWORK = (0, 0, 0, 7, 7, 0, 0)
+FRI_SAT_TO_NEXTWORK = (0, 0, 0, 0, 7, 7, 0)
+NEAREST_MON = (0, -1, -2, -3, 3, 2, 1)
```
[the remark this answers is not in the document]
Comment (KJhellico) [reply holidays/constants.py]: Done.

File: holidays/groups/observed.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        if obs_date < dt:
+            observed_label = getattr(self, "observed_label_before", observed_label)
+        names = [name] if name else self.get_list(dt)
+        for name in names:
+            HolidayBase._add_holiday(
+                self, self.tr(observed_label) % self.tr(name), obs_date  # type: ignore[arg-type]
+            )
+        return True, obs_date
+
+    def _add_observed(
```
[the remark this answers is not in the document]
Comment (KJhellico) [reply holidays/groups/observed.py]: Done.

File: holidays/countries/guatemala.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arkid15r) [reply holidays/countries/guatemala.py]: > "British" `NEAREST_MON` (used in Canada and New Zealand) differs from this: here, Thursday is moved to next Monday, and there, it's moved to previous Monday. :))

Yeah, that's why "be based on" :) The idea was to just update the difference. Anyway, it's not an issue anymore.

File: holidays/groups/observed.py
[the remark this answers is not in the document]
Comment (arkid15r) [reply holidays/groups/observed.py:1]: I took care of this, let me know whether it looks good to you

File: holidays/countries/ecuador.py [hunk 9]
```diff
@@ -27,51 +33,42 @@ class Ecuador(HolidayBase, ChristianHolidays, InternationalHolidays):
...
+        )
 
     def _populate(self, year):
         super()._populate(year)
 
         # New Year's Day.
         name = self.tr("Año Nuevo")
-        self._add_observed(self._add_new_years_day(name), weekend_only=True)
+        self._add_observed(
+            self._add_new_years_day(name), rule=SAT_TO_PREV_WORKDAY + SUN_TO_NEXT_MON
```
Comment (KJhellico) [inline holidays/countries/ecuador.py:67]:
```suggestion
            self._add_new_years_day(name), rule=SAT_TO_PREV_FRI + SUN_TO_NEXT_MON
```

File: holidays/countries/ecuador.py [hunk 10]
```diff
@@ -93,14 +90,23 @@ def _populate(self, year):
...
+        self._add_observed(
+            # Independence of Cuenca.
+            self._add_holiday_nov_3(tr("Independencia de Cuenca")),
+            rule=WED_THU_TO_NEXT_FRI + SUN_TO_NEXT_MON,  # Not observed the previous day.
+        )
+
+        self._add_observed(
+            # Christmas Day.
+            self._add_christmas_day(tr("Día de Navidad")),
+            rule=SAT_TO_PREV_WORKDAY + SUN_TO_NEXT_MON,
```
Comment (KJhellico) [inline holidays/countries/ecuador.py:108]:
```suggestion
            rule=SAT_TO_PREV_FRI + SUN_TO_NEXT_MON,
```

File: holidays/countries/ecuador.py [hunk 11]
```diff
@@ -9,16 +9,22 @@
...
-
-class Ecuador(HolidayBase, ChristianHolidays, InternationalHolidays):
+from holidays.observed_holiday_base import (
+    ObservedHolidayBase,
+    TUE_TO_PREV_MON,
+    WED_TO_NEXT_FRI,
+    SAT_TO_PREV_FRI,
+    SUN_TO_NEXT_MON,
+    WED_THU_TO_NEXT_FRI,
+    SAT_TO_PREV_WORKDAY,
```
Comment (KJhellico) [inline holidays/countries/ecuador.py:23]:
```suggestion
```

File: holidays/observed_holiday_base.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+WED_THU_TO_NEXT_FRI = ObservedRule({WED: +2, THU: +1})
+
+THU_FRI_TO_NEXT_MON = ObservedRule({THU: +4, FRI: +3})
+THU_FRI_TO_NEXT_WORKDAY = ObservedRule({THU: +7, FRI: +7})
+THU_FRI_SUN_TO_NEXT_MON = ObservedRule({THU: +4, FRI: +3, SUN: +1})
+
+FRI_SAT_TO_NEXT_WORKDAY = ObservedRule({FRI: +7, SAT: +7})
+FRI_SUN_TO_NEXT_MON = ObservedRule({FRI: +3, SUN: +1})
+FRI_SUN_TO_NEXT_SAT_MON = ObservedRule({FRI: +1, SUN: +1})
```
[the remark this answers is not in the document]
Comment (arkid15r) [reply holidays/observed_holiday_base.py:80]: I had done exactly this initially but then switched to the current implementation to save an extra call and keep it looking consistent with `ObserverdRule(...)` syntax

File: holidays/countries/kazakhstan.py [hunk 13]
```diff
@@ -94,19 +95,12 @@ def _populate(self, year):
         if 2002 <= year <= 2021:
             dts_observed.add(self._add_holiday_dec_17(name))
 
+        self._populate_observed(dts_observed)
```
[the remark this answers is not in the document]
Comment (arkid15r) [reply holidays/countries/kazakhstan.py]: Fixed

File: holidays/countries/azerbaijan.py [hunk 14]
```diff
@@ -10,48 +10,40 @@
...
     country = "AZ"
+    observed_label = "%s (Observed)"
 
     def __init__(self, *args, **kwargs):
         InternationalHolidays.__init__(self)
         IslamicHolidays.__init__(self, calendar=AzerbaijanIslamicCalendar())
-        super().__init__(*args, **kwargs)
+        super().__init__(
+            observed_rule=SAT_SUN_TO_NEXT_WORKDAY, observed_since=2006, *args, **kwargs
+        )
```
[the remark this answers is not in the document]
Comment (arkid15r) [reply holidays/countries/azerbaijan.py:38]: Done

File: holidays/countries/ecuador.py (hunk 9 above)
Comment (arkid15r) [reply holidays/countries/ecuador.py]: 👍 good chatch

**Review by arkid15r — COMMENTED** [review]
PTAL

**Review by KJhellico — COMMENTED** [review]
LGTM!

**Review by arkid15r — APPROVED** [review]
Let's merge it!

**Pull request merged**